server/src/main/java/io/druid/server/ClientInfoResource.java

```diff
@@ -83,13 +100,83 @@ public ClientInfoResource(
   @Produces(MediaType.APPLICATION_JSON)
   public Map<String, Object> getDatasource(
       @PathParam("dataSourceName") String dataSourceName,
-      @QueryParam("interval") String interval
+      @QueryParam("interval") String interval,
+      @QueryParam("full") String full
   )
   {
-    return ImmutableMap.<String, Object>of(
-        "dimensions", getDatasourceDimensions(dataSourceName, interval),
-        "metrics", getDatasourceMetrics(dataSourceName, interval)
-    );
+    if(full == null) {
+      return ImmutableMap.<String, Object>of(
+          KEY_DIMENSIONS, getDatasourceDimensions(dataSourceName, interval),
+          KEY_METRICS, getDatasourceMetrics(dataSourceName, interval)
+      );
+    }
+
+    Interval theInterval;
+    if (interval == null || interval.isEmpty()) {
+      DateTime now = new DateTime();
+      theInterval = new Interval(now.minusMillis(SEGMENT_HISTORY_MILLIS), now);
+    } else {
+      theInterval = new Interval(interval);
+    }
+
+    TimelineLookup<String, ServerSelector> timeline = timelineServerView.getTimeline(new TableDataSource(dataSourceName));
+    Iterable<TimelineObjectHolder<String, ServerSelector>> serversLookup = timeline != null ? timeline.lookup(theInterval) : null;
+    if(serversLookup == null) {
+      return Collections.EMPTY_MAP;
+    }
+    Map<Interval,Object> servedIntervals = new TreeMap<>(new Comparator<Interval>()
+    {
+      @Override
+      public int compare(Interval o1, Interval o2)
+      {
+        if(o1.equals(o2) || o1.overlaps(o2)) {
+          return 0;
+        } else {
+          return o1.isBefore(o2) ? -1 : 1;
+        }
+      }
+    });
+
+    for (TimelineObjectHolder<String, ServerSelector> holder : serversLookup) {
+      servedIntervals.put(holder.getInterval(), ImmutableMap.of(KEY_DIMENSIONS, Sets.newHashSet(), KEY_METRICS, Sets.newHashSet()));
+    }
+
+    List<DataSegment> segments = getSegmentsForDatasources().get(dataSourceName);
+    if (segments == null || segments.isEmpty()) {
+      //Note: is this check really required? when there are segments in the timeline, there
+      //must be segments in the inventory as well.
+      return Collections.EMPTY_MAP;
+    }
+
+    for (DataSegment segment : segments) {
```

Comment:
1st for loop to go through all the TimelineObjectHolders is required to get list of intervals being served, which is not necessarily same as union of intervals of List<DataSegment> segments.
Let say, for a segment granularity duration, there were 10 partitions and only 9 are being served then I want to consider that interval as not being served and discard those 9 segments. (This is in line with broker behavior on query response, it would discard those 9 segments also).


Now, I imagine, there could be a possibility to combine the loops at line 151 and 163... but that probably requires List<DataSegment> segments at line 144 be "sorted" according to interval which is not necessarily true.